domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/PerformanceMetricsController.kt

```diff
@@ -0,0 +1,178 @@
+    }
+  }
+
+  /** Returns a metric log containing relevant data for metric log reporting. */
+  private fun createMetricLog(
+    timestamp: Long,
+    priority: Priority,
+    currentScreen: OppiaMetricLog.CurrentScreen,
+    loggableMetric: OppiaMetricLog.LoggableMetric
+  ): OppiaMetricLog {
+    return OppiaMetricLog.newBuilder().apply {
+      this.timestampMillis = timestamp
+      this.priority = priority
+      this.currentScreen = currentScreen
+      this.loggableMetric = loggableMetric
+      this.isAppInForeground = performanceMetricsUtils.isAppInForeground()
+      this.storageTier = performanceMetricsUtils.getDeviceStorageTier()
+      this.memoryTier = performanceMetricsUtils.getDeviceMemoryTier()
+    }.build()
+  }
+
+  /**
+   * Returns the index of the least recent event from the existing store on the basis of recency and
+   * priority.
+   *
+   * At first, it checks the index of the least recent event which has LOW priority. If that
+   * returns null, then it checks the index of the least recent event which has MEDIUM priority. If
+   * that returns null, then it checks the index of the least recent event regardless of the
+   * priority is returned.
+   */
+  private fun getLeastRecentMetricLogIndex(oppiaMetricLogs: OppiaMetricLogs): Int? =
+    oppiaMetricLogs.oppiaMetricLogList.withIndex()
+      .filter { it.value.priority == Priority.LOW_PRIORITY }
+      .minByOrNull { it.value.timestampMillis }?.index ?: getLeastRecentMediumPriorityEventIndex(
+      oppiaMetricLogs
+    )
+
+  /**
+   * Returns the index of the least recent event from the existing store on the basis of recency and
+   * priority.
+   *
+   * At first, it checks the index of the least recent event which has LOW priority. If that
+   * returns null, then it checks the index of the least recent event regardless of the
+   * priority is returned.
+   */
+  private fun getLeastRecentMediumPriorityEventIndex(oppiaMetricLogs: OppiaMetricLogs): Int? =
+    oppiaMetricLogs.oppiaMetricLogList.withIndex()
+      .filter { it.value.priority == Priority.MEDIUM_PRIORITY }
+      .minByOrNull { it.value.timestampMillis }?.index ?: getLeastRecentGeneralEventIndex(
+      oppiaMetricLogs
+    )
+
+  /** Returns the index of the least recent event regardless of their priority. */
+  private fun getLeastRecentGeneralEventIndex(oppiaMetricLogs: OppiaMetricLogs): Int? =
+    oppiaMetricLogs.oppiaMetricLogList.withIndex()
+      .minByOrNull { it.value.timestampMillis }?.index
+
+  /** Returns a data provider for log reports that have been recorded for upload. */
+  fun getMetricLogStore(): DataProvider<OppiaMetricLogs> = metricLogStore
+
+  /**
+   * Returns a list of metric log reports that have been recorded for upload.
+   *
+   * As we are using the await call on the deferred output of readDataAsync, the failure case would
+   * be caught and it'll throw an error.
+   */
+  suspend fun getMetricLogStoreList(): MutableList<OppiaMetricLog> {
```

Comment:
The returned list is actually immutable, so we should just be returning 'List' here.

Reply:
done.